lazyflow/graph.py

```diff
@@ -80,16 +79,43 @@
 from lazyflow.metaDict import MetaDict
 
 
-class Graph(object):
+class Graph:
     """
     A Graph instance is shared by all connected operators and contains any
     bookkeeping or globally accessible state needed by all operators/slots in the graph.
     """
 
+    class Transaction:
```

Comment:
Maybe we could support nested transactions just by keeping a list of lists?

```python
def __init__(self):
    self._deferred_callbacks = []

def on_exit(self, fn):
    assert self._deferred_callbacks
    cbs = self._deferred_callbacks[-1]
    if fn not in cbs:
        cbs.append(fn)        

def __enter__(self):
    self._deferred_callbacks.append([])

def __exit__(self, ...):
    for cb in self._deferred_callbacks.pop():
        cb()
```

Reply:
Do you have a use-case in mind for nested transactions?

Reply:
> Do you have a use-case in mind for nested transactions?

More composable functions/methods:

```python
with op.transaction:
    op.setValue("spam")
    do_stuff(op)

# 1000 lines later...

def do_stuff(op):
    with op.transaction:
        op.setValue("eggs")
```

Although I am not sure that this would be a common use case. However, this seems to be a low-hanging fruit.

Reply:
I would defer implementing nested transactions.
For example what do you except on exit from a nested transaction? Should I be just reentrant transaction on the graph?